### PR TITLE
Add custom retry logic for cloud run

### LINKED ISSUE #399: Status fields should be seen as outputs
It seems like the output values aren't being populated. Subsequent resources cannot use the outputs of the service even with an explicit dependency. I need to do two separate deployments and run Pulumi refresh in between in order to use the output values.

```
const version = "v3";  // increment this to force an update
const service = new google.run.v1.Service("test-service", {
    apiVersion: "serving.knative.dev/v1",
    kind: "Service",
    project: 'projectidxxx',
    location: 'us-east4',
    metadata: {
        labels: {"version": version},
        name: "test-service"
    },
    spec: {
        template: {
            spec: {
                containers: [{image: "gcr.io/cloudrun/hello"}],
            },
        },
    },
});
export const exportedUrl = service.status.url;
const scheduler = new google.cloudscheduler.v1.Job('schedule_job', {
    project: 'projectidxxx',
    location: 'us-east4',
    name: `projects/projectidxxx/locations/us-east4/jobs/schedule-job-test`,
    schedule: "0 6 * * *",
    attemptDeadline: "1800s",
    httpTarget: {
        httpMethod: 'GET',
        uri: service.status.url
    }
}, {
    dependsOn: service
});
```

```
colesiegel@TR-C02VD7EGHTD6 repro % pulumi up --non-interactive --skip-preview --yes
Updating (test-bug-repro):

 +  pulumi:pulumi:Stack bug-repro-test-bug-repro creating 
 +  google-native:run/v1:Service test-service creating 
 +  google-native:run/v1:Service test-service created 
 +  google-native:cloudscheduler/v1:Job schedule_job creating 
 +  google-native:cloudscheduler/v1:Job schedule_job creating error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 +  google-native:cloudscheduler/v1:Job schedule_job **creating failed** error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 +  pulumi:pulumi:Stack bug-repro-test-bug-repro creating error: update failed
 +  pulumi:pulumi:Stack bug-repro-test-bug-repro **creating failed** 1 error
 
Diagnostics:
  pulumi:pulumi:Stack (bug-repro-test-bug-repro):
    error: update failed
 
  google-native:cloudscheduler/v1:Job (schedule_job):
    error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 
Resources:
    + 2 created

Duration: 16s

colesiegel@TR-C02VD7EGHTD6 repro % pulumi up --non-interactive --skip-preview --yes
Updating (test-bug-repro):

    pulumi:pulumi:Stack bug-repro-test-bug-repro running 
    google-native:run/v1:Service test-service  
 +  google-native:cloudscheduler/v1:Job schedule_job creating 
 +  google-native:cloudscheduler/v1:Job schedule_job creating error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 +  google-native:cloudscheduler/v1:Job schedule_job **creating failed** error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
    pulumi:pulumi:Stack bug-repro-test-bug-repro running error: update failed
    pulumi:pulumi:Stack bug-repro-test-bug-repro **failed** 1 error
 
Diagnostics:
  google-native:cloudscheduler/v1:Job (schedule_job):
    error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 
  pulumi:pulumi:Stack (bug-repro-test-bug-repro):
    error: update failed
 
Resources:
    2 unchanged

Duration: 5s

colesiegel@TR-C02VD7EGHTD6 repro % pulumi refresh
Previewing refresh (test-bug-repro):
     Type                             Name                      Plan       Info
     pulumi:pulumi:Stack              bug-repro-test-bug-repro             
 ~   └─ google-native:run/v1:Service  test-service              update     [diff: ~metadata,status]
 
Resources:
    ~ 1 to update
    1 unchanged

Do you want to perform this refresh?
No resources will be modified as part of this refresh; just your stack's state will be. yes
Refreshing (test-bug-repro):
     Type                             Name                      Status      Info
     pulumi:pulumi:Stack              bug-repro-test-bug-repro              
 ~   └─ google-native:run/v1:Service  test-service              updated     [diff: ~metadata,status]
 
Resources:
    ~ 1 updated
    1 unchanged

Duration: 1s

colesiegel@TR-C02VD7EGHTD6 repro % pulumi up --non-interactive --skip-preview --yes
Updating (test-bug-repro):

    pulumi:pulumi:Stack bug-repro-test-bug-repro running 
 ~  google-native:run/v1:Service test-service updating [diff: -status~metadata]
 ~  google-native:run/v1:Service test-service updated [diff: -status~metadata]
 +  google-native:cloudscheduler/v1:Job schedule_job creating 
 +  google-native:cloudscheduler/v1:Job schedule_job created 
    pulumi:pulumi:Stack bug-repro-test-bug-repro  
 
Outputs:
  + exportedUrl: "https://test-service-jk33cxl7ga-uk.a.run.app"

Resources:
    + 1 created
    ~ 1 updated
    2 changes. 1 unchanged

Duration: 6s
```

Even if I specify `--refresh` it fails on the first deployment:

```
colesiegel@TR-C02VD7EGHTD6 repro % pulumi up --non-interactive --skip-preview --yes --refresh
Updating (test-bug-repro):

 +  pulumi:pulumi:Stack bug-repro-test-bug-repro creating 
 +  google-native:run/v1:Service test-service creating 
 +  google-native:run/v1:Service test-service created 
 +  google-native:cloudscheduler/v1:Job schedule_job creating 
 +  google-native:cloudscheduler/v1:Job schedule_job creating error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 +  google-native:cloudscheduler/v1:Job schedule_job **creating failed** error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 +  pulumi:pulumi:Stack bug-repro-test-bug-repro creating error: update failed
 +  pulumi:pulumi:Stack bug-repro-test-bug-repro **creating failed** 1 error
 
Diagnostics:
  pulumi:pulumi:Stack (bug-repro-test-bug-repro):
    error: update failed
 
  google-native:cloudscheduler/v1:Job (schedule_job):
    error: error sending request: googleapi: Error 400: URL cannot be empty.: "https://cloudscheduler.googleapis.com/v1/projects/projectidxxx/locations/us-east4/jobs" map[attemptDeadline:1800s httpTarget:map[httpMethod:GET] location:us-east4 name:projects/projectidxxx/locations/us-east4/jobs/schedule-job-test project:projectidxxx schedule:0 6 * * *]
 
Resources:
    + 2 created

Duration: 7s
```

_Originally posted by @ColeSiegelTR in https://github.com/pulumi/pulumi-google-native/issues/256#issuecomment-995052951_

### PR DESCRIPTION
**Pull request by viveklak**
Fixes #399.

Adds a heuristic to determine output-ty status fields from types used as inputs and outputs where these were previously being construed as required inputs due to missing metadata.

Also adds some machinery to override the metadata for some resources to make one-of fixes easier.

### CODE REVIEW

**Comment by viveklak** [comment]
> ### Does the PR have any schema changes?
> Resource "google-native:dns/v2:ManagedZone" missing input "id"
> Resource "google-native:dns/v2:ManagedZone" missing input "nameServers"
> Resource "google-native:dns/v1beta2:Change" missing input "startTime"
> Resource "google-native:dns/v1beta2:Change" missing input "id"
> Resource "google-native:dns/v1beta2:Change" missing input "status"
> Resource "google-native:sqladmin/v1:BackupRun" missing input "status"
> Resource "google-native:run/v1:Service" missing input "status"
> Resource "google-native:run/v1:DomainMapping" missing input "status"
> Resource "google-native:accesscontextmanager/v1beta:ServicePerimeter" missing input "status"
> Resource "google-native:dns/v1beta2:ResponsePolicy" missing input "id"
> Resource "google-native:deploymentmanager/alpha:CompositeType" missing input "status"
> Resource "google-native:compute/v1:NodeGroup" missing input "status"
> Resource "google-native:dns/v1beta2:ManagedZone" missing input "id"
> Resource "google-native:dns/v1beta2:ManagedZone" missing input "nameServers"
> Resource "google-native:compute/alpha:NodeGroup" missing input "status"
> Resource "google-native:dns/v2:Policy" missing input "id"
> Resource "google-native:dns/v1:Policy" missing input "id"
> Resource "google-native:run/v1alpha1:Job" missing input "status"
> Resource "google-native:accesscontextmanager/v1:ServicePerimeter" missing input "status"
> Resource "google-native:dns/v1beta2:Policy" missing input "id"
> Resource "google-native:deploymentmanager/v2beta:CompositeType" missing input "status"
> Resource "google-native:servicemanagement/v1:Rollout" missing input "status"
> Resource "google-native:sqladmin/v1beta4:BackupRun" missing input "status"
> Resource "google-native:dns/v1:Change" missing input "status"
> Resource "google-native:dns/v1:Change" missing input "id"
> Resource "google-native:dns/v1:Change" missing input "startTime"
> Resource "google-native:compute/beta:NodeGroup" missing input "status"
> Resource "google-native:dns/v1:ManagedZone" missing input "id"
> Resource "google-native:dns/v1:ManagedZone" missing input "nameServers"

These are output-only fields that should not have been modeled as inputs.

> Type "google-native:dns/v2:ChangeStatus" missing
> Type "google-native:dns/v1:ChangeStatus" missing
> Type "google-native:dns/v1beta2:ChangeStatus" missing

These are affected by https://github.com/pulumi/pulumi-google-native/issues/413. I will open a separate PR to introduce these types back (along with other enums which are currently excluded).

File: provider/pkg/resources/resources.go [hunk 1]
```diff
@@ -119,10 +120,23 @@ type ResourceAutoname struct {
 	ValidationRegex string `json:"validationRegex,omitempty"`
 }
 
+type PollingStrategy string
+
+const (
+	DefaultPoll       = PollingStrategy("")
+	KNativeStatusPoll = PollingStrategy("KNativeStatusPoll")
```
Comment (viveklak) [inline provider/pkg/resources/resources.go:127]: We will likely have to support additional custom polling mechanisms here. See https://github.com/GoogleCloudPlatform/declarative-resource-client-library/tree/main/dcl/operations

**Comment by viveklak** [comment]
Reviewer note:
The primary change is in https://github.com/pulumi/pulumi-google-native/pull/411/files#diff-04c663ce0ec07699115794969a983af85f7def6794d0a9170e1c1645c6bf2568 (`poll.go`) where a new custom polling function for knative (cloud run) resources is added. Changes in [`overrides.go`](https://github.com/pulumi/pulumi-google-native/pull/411/files#diff-85d5fad1cb24cc2a4df0556778701885ceb8de1c07b5ea233659933f2d4d8381) and [`schema.go`](https://github.com/pulumi/pulumi-google-native/pull/411/files#diff-4b4c0cf25563429565383180f98332b1dd806ba3fc7b7cd8473ba51435199667) are to add a marker in the metadata for relevant resources for using the custom polling mechanism. I also fixed a couple of issues where output-ty fields (e.g. status) were being marked as input types in the schema causing await issues. This is probably the hackiest part of the PR. This is also the reason why you see schema/sdk changes in this PR.

File: provider/pkg/provider/poll.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+				return false, fmt.Errorf(`resource is in failed state "Ready:False", message: %s`,
+					condition.Message)
+			}
+		}
+	}
+	return false, nil
+}
+
+func extractMetadata(res map[string]interface{}) (map[string]interface{}, error) {
+	// 1. Get generation from metadata
```
Comment (jkodroff) [inline provider/pkg/provider/poll.go:78]: Would like a godoc comment on the func, and this comment removed.

File: provider/pkg/resources/resources.go [hunk 3]
```diff
@@ -119,10 +120,23 @@ type ResourceAutoname struct {
 	ValidationRegex string `json:"validationRegex,omitempty"`
 }
 
+type PollingStrategy string
+
+const (
+	DefaultPoll       = PollingStrategy("")
```
Comment (jkodroff) [inline provider/pkg/resources/resources.go:126]: `None`, maybe?

**Review by jkodroff — APPROVED** [review]
Approved with optional suggestions.

**Review by lblackstone — CHANGES_REQUESTED** [review]

File: provider/pkg/gen/schema.go [hunk 4]
```diff
@@ -470,7 +471,18 @@ func (g *packageGenerator) genResource(typeName string, dd discoveryDocumentReso
 			return err
 		}
 
+		isOutput := func(desc string) bool {
+			lowerDesc := strings.ToLower(desc)
+			return isReadOnly(lowerDesc) ||
+				!(strings.Contains(lowerDesc, "values include") ||
+					strings.Contains(lowerDesc, "value must be specified"))
+		}
```
[the remark this answers is not in the document]
Comment (viveklak) [reply provider/pkg/gen/schema.go:479]: Not sure. I think i tried to roll this into `isReadOnly` and it was a bit too broad? Lets keep it's usage a bit constrained for now.

File: provider/pkg/googleclient/http.go [hunk 5]
```diff
@@ -132,10 +132,8 @@ func (c *GoogleClient) RequestWithTimeout(
 		return nil, err
 	}
 
-	// TODO: Re-enable response dump when https://github.com/golang/go/issues/49366 fix
-	// propagates to our current Go version. This fix will be in Go 1.18.
-	debugResp, _ := httputil.DumpResponse(res, false)
-	logging.V(9).Infof(responseFormat, res.Request.Method, res.Request.URL, prettyPrintJSONLines(debugResp))
```
[the remark this answers is not in the document]
Comment (viveklak) [reply provider/pkg/googleclient/http.go:138]: I believe this has been backported to 1.17 as well: https://github.com/golang/go/issues/49559

File: examples/cloudrun-ts/step1/index.ts [hunk 6]
```diff
@@ -27,6 +27,17 @@ const service = new google.run.v1.Service("service", {
     },
 });
 
+export const exportedUrl = service.status.url;
+const scheduler = new google.cloudscheduler.v1.Job('schedule_job', {
+    location: 'us-east4',
```
[the remark this answers is not in the document]
Comment (viveklak) [reply examples/cloudrun-ts/step1/index.ts]: Removed, thanks for the catch.

File: provider/pkg/provider/poll.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+	// 3. Check if observed and actual generation match. If not, continue to poll
+	if int(s.Status.ObservedGeneration) != generation {
+		return false, nil
+	}
+
+	// 4. Check conditions to ensure resource is ready.
+	for _, condition := range s.Status.Conditions {
+		if condition.Type == readyStatusType {
+			log.Printf("[DEBUG] checking KnativeStatus Ready condition %s: %s", condition.Status, condition.Message)
```
[the remark this answers is not in the document]
Comment (viveklak) [reply provider/pkg/provider/poll.go]: Yup - fixed!

**Review by lblackstone — APPROVED** [review]

**Pull request merged**